### PR TITLE
Add masonry layout grid

### LINKED ISSUE #378: 🐛[Bug] All PostCards have the same height - dynamic height required
> [ulfgebhardt](https://github.com/ulfgebhardt) **Authored by [ulfgebhardt](https://github.com/ulfgebhardt)**
_<time datetime="2019-04-03T12:24:52Z" title="Wednesday, April 3rd 2019, 2:24:52 pm +02:00">Apr 3, 2019</time>_
_Closed <time datetime="2019-08-20T22:42:49Z" title="Wednesday, August 21st 2019, 12:42:49 am +02:00">Aug 21, 2019</time>_
---

### All PostCards have the same height - dynamic height required

max height must be defined

image

image

### Steps to reproduce the behavior
1. Login
2. See Cards
3. All same Height
4. ...
5. Profit

### Expected behavior

image

### Version & Environment
 Type: [Desktop]      
 - OS: [Windows]      
 - Browser: [Chrome Beta] 
 - Version https://github.com/Human-Connection/Human-Connection/commit/e4cfa2403a457032c10b449abeaf3082e1863111  

### Additional Context

Keyword: masonry grid

### PR DESCRIPTION
**Pull request by Tirokk**
> [alina-beck](https://github.com/alina-beck) **Authored by [alina-beck](https://github.com/alina-beck)**
_<time datetime="2019-08-13T15:54:03Z" title="Tuesday, August 13th 2019, 5:54:03 pm +02:00">Aug 13, 2019</time>_
_Merged <time datetime="2019-08-20T22:42:49Z" title="Wednesday, August 21st 2019, 12:42:49 am +02:00">Aug 21, 2019</time>_
---

## 🍰 Pullrequest
Use CSS grid to create a masonry layout – allowing posts to have different heights.

So instead of this:
Screenshot 2019-08-13 at 16 42 47

We should now see this:
Screenshot 2019-08-13 at 16 49 48

### Issues
- fixes #378
- relates #372
- waiting for PR: https://github.com/Human-Connection/Nitro-Styleguide/pull/131

### ToDo
- [x] update styleguide dependency
- [x] use styleguide components in project

### CODE REVIEW

File: webapp/components/PostCard/index.vue [hunk 1]
```diff
@@ -1,69 +1,71 @@
...
+      <hc-ribbon :text="$t('post.name')" />
+    </div>
+    <ds-space margin-bottom="small" />
+    <!-- Post Title -->
+    <ds-heading tag="h3" no-margin>{{ post.title }}</ds-heading>
+    <ds-space margin-bottom="small" />
+    <!-- Post Content Excerpt -->
+    <!-- eslint-disable vue/no-v-html -->
+    <!-- TODO: replace editor content with tiptap render view -->
+    <div class="hc-editor-content" v-html="excerpt" />
```
Comment (Tirokk) [inline webapp/components/PostCard/index.vue:28]: > [roschaefer](https://github.com/roschaefer) **Authored by [roschaefer](https://github.com/roschaefer)**
_<time datetime="2019-08-14T22:36:05Z" title="Thursday, August 15th 2019, 12:36:05 am +02:00">Aug 15, 2019</time>_

---
> **Outdated (history rewrite)** - original diff
---
```diff
@@ -123,6 +132,25 @@ export default {
         this.$toast.error(err.message)
       }
     },
+    calculateItemHeight() {
+      const grid = document.querySelector('.masonry-grid')
```


We have to check if globals like `document` or `window` are available. Our tests are failing because of it.

Comment (Tirokk) [reply webapp/components/PostCard/index.vue:28]: > [roschaefer](https://github.com/roschaefer) **Authored by [roschaefer](https://github.com/roschaefer)**
_<time datetime="2019-08-14T22:38:57Z" title="Thursday, August 15th 2019, 12:38:57 am +02:00">Aug 15, 2019</time>_

---


Also I would find it cleaner in this case if the parent would interact with its children instead of the other way round. According to this diagram:
https://stackoverflow.com/questions/52433712/vue-js-component-lifecycle-for-children
vue lifecycle hooks for parent components
You could create a masonry component that iterates through all children with a particular class and sets the height. I don't know if this might work or if this still works if the list gets updated.. but worth a try.

**Comment by Tirokk** [comment]
> [roschaefer](https://github.com/roschaefer) **Authored by [roschaefer](https://github.com/roschaefer)**
_<time datetime="2019-08-15T09:04:24Z" title="Thursday, August 15th 2019, 11:04:24 am +02:00">Aug 15, 2019</time>_

---


@alina-beck I rebased your PR - you will need to `git fetch && git reset origin/add-masonry-grid --hard`

**Comment by Tirokk** [comment]
> [alina-beck](https://github.com/alina-beck) **Authored by [alina-beck](https://github.com/alina-beck)**
_<time datetime="2019-08-15T09:36:08Z" title="Thursday, August 15th 2019, 11:36:08 am +02:00">Aug 15, 2019</time>_

---


@roschaefer this is still a work in progress – I need the [styleguide PR](https://github.com/Human-Connection/Nitro-Styleguide/pull/131) merged and released, then I can finish this one. I still have a few plans. :wink: But thank you for your suggestions! The diagram is really helpful. 😍 

May I ask why you are force-pushing and rebasing instead of merging in master? I've only ever done it locally and thought it wouldn't be a good idea to do it on the remote repo... Now I'm curious to hear your reasons because I've seen you do it quite often!

File: webapp/components/PostCard/index.vue [hunk 2]
```diff
@@ -1,69 +1,71 @@
...
+      </no-ssr>
+      <hc-ribbon :text="$t('post.name')" />
+    </div>
+    <ds-space margin-bottom="small" />
+    <!-- Post Title -->
+    <ds-heading tag="h3" no-margin>{{ post.title }}</ds-heading>
+    <ds-space margin-bottom="small" />
+    <!-- Post Content Excerpt -->
+    <!-- eslint-disable vue/no-v-html -->
+    <!-- TODO: replace editor content with tiptap render view -->
```
Comment (Tirokk) [inline webapp/components/PostCard/index.vue:27]: > [roschaefer](https://github.com/roschaefer) **Authored by [roschaefer](https://github.com/roschaefer)**
_<time datetime="2019-08-15T22:35:24Z" title="Friday, August 16th 2019, 12:35:24 am +02:00">Aug 16, 2019</time>_

---
> **Outdated (history rewrite)** - original diff
---
```diff
@@ -123,6 +132,26 @@ export default {
         this.$toast.error(err.message)
       }
     },
+    calculateItemHeight() {
```


@alina-beck do you still need this? The build server is crashing during frontend testing and it's trying to set some `style` attribute of `null`.

File: webapp/components/MasonryGrid/MasonryGridItem.vue [hunk 3]
```diff
@@ -0,0 +1,39 @@
...
+<script>
+export default {
+  data() {
+    return {
+      rowSpan: 10,
+    }
+  },
+  methods: {
+    calculateItemHeight() {
+      this.$parent.$emit('calculating-item-height')
```
Comment (Tirokk) [inline webapp/components/MasonryGrid/MasonryGridItem.vue:16]: > [mattwr18](https://github.com/mattwr18) **Authored by [mattwr18](https://github.com/mattwr18)**
_<time datetime="2019-08-20T07:03:30Z" title="Tuesday, August 20th 2019, 9:03:30 am +02:00">Aug 20, 2019</time>_

---
> **Outdated (history rewrite)** - original diff
---
```diff
@@ -0,0 +1,38 @@
+<template>
+  <ds-grid-item :rowSpan="rowSpan">
+    <slot></slot>
+  </ds-grid-item>
+</template>
+
+<script>
+export default {
+  data() {
+    return {
+      rowSpan: 10,
+    }
+  },
+  methods: {
+    calculateItemHeight() {
+      this.$parent.$emit('calculating-item-height')
```


I was curious what the difference between `this.$emit` and `this.$parent.$emit` was, and found this SO post.
As far as I understand it, this is the way to tell the parent component to emit an event to its parent? so like emit an event to a grandparent component.
Is that correct @alina-beck?

File: webapp/components/MasonryGrid/MasonryGridItem.vue [hunk 4]
```diff
@@ -0,0 +1,39 @@
...
+      })
+    },
+  },
+  mounted() {
+    const image = this.$el.querySelector('img')
+    if (image) {
+      image.onload = () => this.calculateItemHeight()
+    } else {
+      // use timeout to make sure layout is set up before calculation
+      setTimeout(() => this.calculateItemHeight(), 0)
```
Comment (Tirokk) [inline webapp/components/MasonryGrid/MasonryGridItem.vue:35]: > [mattwr18](https://github.com/mattwr18) **Authored by [mattwr18](https://github.com/mattwr18)**
_<time datetime="2019-08-20T07:05:01Z" title="Tuesday, August 20th 2019, 9:05:01 am +02:00">Aug 20, 2019</time>_

---
> **Outdated (history rewrite)** - original diff
---
```diff
@@ -0,0 +1,38 @@
+<template>
+  <ds-grid-item :rowSpan="rowSpan">
+    <slot></slot>
+  </ds-grid-item>
+</template>
+
+<script>
+export default {
+  data() {
+    return {
+      rowSpan: 10,
+    }
+  },
+  methods: {
+    calculateItemHeight() {
+      this.$parent.$emit('calculating-item-height')
+      this.$nextTick(() => {
+        const grid = this.$parent
+        const rowHeight = parseInt(grid.$el.style.gridAutoRows)
+        const rowGap = parseInt(grid.$el.style.gridRowGap)
+        const itemHeight = this.$el.clientHeight
+
+        this.rowSpan = Math.ceil((itemHeight + rowGap) / (rowHeight + rowGap))
+        this.$parent.$emit('finished-calculating-item-height')
+      })
+    },
+  },
+  mounted() {
+    const image = this.$el.querySelector('img')
+    if (image) {
+      image.onload = () => this.calculateItemHeight()
+    } else {
+      // use timeout to make sure layout is set up before calculation
+      setTimeout(() => this.calculateItemHeight(), 0)
```


I like that you added a comment to explain this, but if the timeout is `0`, I'm a little confused what this does

File: webapp/pages/index.vue [hunk 5]
```diff
@@ -1,10 +1,10 @@
 <template>
   <div>
-    <ds-flex :width="{ base: '100%' }" gutter="base">
-      <ds-flex-item>
+    <masonry-grid>
+      <ds-grid-item v-show="hashtag" :row-span="2" column-span="fullWidth">
```
Comment (Tirokk) [inline webapp/pages/index.vue:4]: > [mattwr18](https://github.com/mattwr18) **Authored by [mattwr18](https://github.com/mattwr18)**
_<time datetime="2019-08-20T07:07:21Z" title="Tuesday, August 20th 2019, 9:07:21 am +02:00">Aug 20, 2019</time>_

---
> **Outdated (history rewrite)** - original diff
---
```diff
@@ -1,10 +1,10 @@
 <template>
   <div>
-    <ds-flex :width="{ base: '100%' }" gutter="base">
-      <ds-flex-item>
+    <masonry-grid>
+      <ds-grid-item v-show="hashtag" :row-span="2" column-span="fullWidth">
```


I was wondering where the `hashtag` was :smiley_cat:

File: webapp/components/MasonryGrid/MasonryGridItem.vue (hunk 3 above)
Comment (Tirokk) [reply webapp/components/MasonryGrid/MasonryGridItem.vue:16]: > [alina-beck](https://github.com/alina-beck) **Authored by [alina-beck](https://github.com/alina-beck)**
_<time datetime="2019-08-20T09:12:14Z" title="Tuesday, August 20th 2019, 11:12:14 am +02:00">Aug 20, 2019</time>_

---


Yes, and the reason we have to do it here is that this component sits within a slot – I found the solution in [this comment](https://github.com/vuejs/vue/issues/4332#issuecomment-263444492) and it's working! But hard to test... 😕

File: webapp/components/MasonryGrid/MasonryGridItem.vue (hunk 4 above)
Comment (Tirokk) [reply webapp/components/MasonryGrid/MasonryGridItem.vue:35]: > [alina-beck](https://github.com/alina-beck) **Authored by [alina-beck](https://github.com/alina-beck)**
_<time datetime="2019-08-20T09:24:50Z" title="Tuesday, August 20th 2019, 11:24:50 am +02:00">Aug 20, 2019</time>_

---


This has to do with the [JavaScript event loop](https://stackoverflow.com/questions/33955650/what-is-settimeout-doing-when-set-to-0-milliseconds/33955673). The timeout – even though it is set to `0` – sends the execution of this function to the back of the queue.

I am not sure which function _exactly_ we are waiting for because I don't know how Vue.js works its magic in the background but I know that without the timeout the layout is messed up. :smile:

**Comment by Tirokk** [comment]
> [alina-beck](https://github.com/alina-beck) **Authored by [alina-beck](https://github.com/alina-beck)**
_<time datetime="2019-08-20T09:36:38Z" title="Tuesday, August 20th 2019, 11:36:38 am +02:00">Aug 20, 2019</time>_

---


@mattwr18 thanks for the review!

I noticed the hashtag issue – and I believe it has nothing to do with this PR. The first thing I did after Rob rebased my commits was test the hashtag feature (because I needed to toggle the `FilterMenu` and wanted to know what was going on) and already then realised that I could click on a hashtag but wouldn't see all the relevant posts...

Update: I'm seeing the same issue on `master`

**Comment by Tirokk** [comment]
> [alina-beck](https://github.com/alina-beck) **Authored by [alina-beck](https://github.com/alina-beck)**
_<time datetime="2019-08-20T14:46:10Z" title="Tuesday, August 20th 2019, 4:46:10 pm +02:00">Aug 20, 2019</time>_

---


@roschaefer Good catch! Thanks to you I found out that 

> CSS Grid Layout initially defined the `grid-row-gap` property. This prefixed property is being replaced by `row-gap`. (See [MDN](https://developer.mozilla.org/en-US/docs/Web/CSS/row-gap))

and am now using the new property (with the old one as a fallback). 😎

File: webapp/components/MasonryGrid/MasonryGridItem.vue (hunk 3 above)
Comment (Tirokk) [reply webapp/components/MasonryGrid/MasonryGridItem.vue:16]: > [roschaefer](https://github.com/roschaefer) **Authored by [roschaefer](https://github.com/roschaefer)**
_<time datetime="2019-08-20T21:41:27Z" title="Tuesday, August 20th 2019, 11:41:27 pm +02:00">Aug 20, 2019</time>_

---


~You could emit again in the parent component
(child)-emit->(parent)-emit->(granddaddy)~ oh no, slots (read the linked issue)

File: webapp/components/MasonryGrid/MasonryGrid.vue [hunk 6]
```diff
@@ -0,0 +1,35 @@
+<template>
+  <ds-grid
+    :min-column-width="300"
+    v-on:calculating-item-height="startCalculation"
```
Comment (Tirokk) [inline webapp/components/MasonryGrid/MasonryGrid.vue:4]: > [roschaefer](https://github.com/roschaefer) **Authored by [roschaefer](https://github.com/roschaefer)**
_<time datetime="2019-08-20T21:55:39Z" title="Tuesday, August 20th 2019, 11:55:39 pm +02:00">Aug 20, 2019</time>_

---
> **Outdated (history rewrite)** - original diff
---
```diff
@@ -0,0 +1,35 @@
+<template>
+  <ds-grid
+    :min-column-width="300"
+    v-on:calculating-item-height="startCalculation"
```


```suggestion
    @calculating-item-height="startCalculation"
```

**Review by Mogge — APPROVED** [review]
Really great stuff @alina-beck!!
I am trying to test it with production data, but I need to pull down all the photos still, so I don't want to hold this up. 
I'm very happy to have the opportunity to learn from your `PR`s!! Excellent!!

One thing I noticed that I'm not sure if is related to changes you made or is a new bug that I hadn't noticed before, the hashtags filter is broken. I'll check on `master` as well, but just curious if you noticed that? 
Screenshot at 2019-08-20 08:57:20
Screenshot at 2019-08-20 08:55:52

**Review by Mogge — CHANGES_REQUESTED** [review]
On archlinux and chromium I see this:
Screenshot - 2019-08-20T143843 832

Here are my specs:
```
robert@e480 ~> chromium --version
Chromium 76.0.3809.100 Arch Linux
robert@e480 ~> cat /proc/version
Linux version 5.2.8-arch1-1-ARCH (builduser@heftig-63215) (gcc version 9.1.0 (GCC)) #1 SMP PREEMPT Fri Aug 9 21:36:07 UTC 2019
```

Same behaviour on Google Chrome but on Firefox it's looking great:
```
robert@e480 ~> google-chrome-stable  --version
Google Chrome 76.0.3809.87
robert@e480 ~> firefox --version
Mozilla Firefox 68.0.2
```

**Review by Mogge — APPROVED** [review]
Screenshot - 2019-08-20T233918 360

The grid now looks amazing on Chromium+Firefox!

finally


I would like to understand a little better why the `MasonryGridItems` must be so clingy. As far as I can see, it all boils down to the `.reset-grid-height` of the parent grid item. So what is the worst thing that could happen if, let's say, not all child components $emit they are done with calculating their heights? It does not really break, or does it?
